File: src/main/java/io/xol/chunkstories/api/particles/ParticleTypeDefinition.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package io.xol.chunkstories.api.particles;

import io.xol.chunkstories.api.content.Content;
import io.xol.chunkstories.api.content.Definition;

public interface ParticleTypeDefinition extends Definition
{
	public String getName();
	
	public Content.ParticlesTypes store();
	
	///** When should we render those particle types */
	//public enum RenderTime {
	//	/** Actually iterated anyways! Use case is light particles, mostly. */
	//	NEVER,
	//	/** Done after the rest of the opaque GBuffer stuff. */
	//	GBUFFER,
	//	FORWARD
	//	;
	//}
	
	public enum ParticleRenderingMode {
		BILLBOARD;
	}
	
	/** Returns the name of the render pass we should use to render those particles. */ 
	public abstract String getRenderPass();
	
	public default String getShaderName()
	{
		return "particles";
	}
	
	/** Returns null or a path to an asset. */
	public abstract String getAlbedoTexture();
	
	/** Returns null or a path to an asset. */
	public abstract String getNormalTexture();

	/** Returns null or a path to an asset. */
	public abstract String getMaterialTexture();
	
	/** Defaults to 1.0f */
	public abstract float getBillboardSize();
}

File: src/main/java/xyz/chunkstories/api/particles/ParticleTypeDefinition.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.particles

import xyz.chunkstories.api.content.Content
import xyz.chunkstories.api.content.json.Json
import xyz.chunkstories.api.content.json.asFloat
import xyz.chunkstories.api.content.json.asString

class ParticleTypeDefinition(val store: Content.ParticlesTypes, val name: String, val properties: Json.Dict) {
    fun store() = store

    /**Returns the name of the render pass we should use to render those particles.  */
    val renderPass: String = properties["renderPass"].asString ?: "forward"// resolveProperty("renderPass", "forward")

    val shaderName: String = properties["shader"].asString ?: "particles"// resolveProperty("shader", "particles")

    /** Returns null or a path to an asset.  */
    val albedoTexture: String = properties["albedoTexture"].asString ?: "res/textures/particle.png"// resolveProperty("albedoTexture", "res/textures/particle.png")

    /** Returns null or a path to an asset.  */
    val normalTexture: String? = properties["normalTexture"].asString // resolveProperty("normalTexture")

    /** Returns null or a path to an asset.  */
    val materialTexture: String? = properties["materialTexture"].asString // resolveProperty("materialTexture")

    /** Defaults to 1.0f  */
    val billboardSize: Float = properties["billboardSize"].asFloat ?: 1.0F
}
